airflow/contrib/operators/bigquery_export_query_results_to_gcs.py

```diff
@@ -0,0 +1,138 @@
+            self,
+            query,
+            destination_cloud_storage_uris,
+            project_id,
+            field_delimiter='\t',
+            bigquery_conn_id='bigquery_default',
+            compression='NONE',
+            print_header=False,
+            export_format='CSV',
+            dataset_id='BigQueryExportQueryResultsToGCS',
+            persist_temp_table=False,
+            delegate_to=None,
+            labels=None,
+            location=None,
+            default_table_expiry_in_ms='3600000',
+            *args,
+            **kwargs):
+        super(BigQueryExportQueryResultsToGCS, self).__init__(*args, **kwargs)
+
+        self.query = query
+        self.destination_cloud_storage_uris = destination_cloud_storage_uris
+        self.field_delimiter = field_delimiter
+        self.project_id = project_id
+        self.bigquery_conn_id = bigquery_conn_id
+        self.compression = compression
+        self.print_header = print_header
+        self.export_format = export_format
+        self.dataset_id = dataset_id + str(uuid4()).replace('-', '')
+        self.temp_table_name = self.project_id + '.%s.%s' % (self.dataset_id, str(uuid4()).replace('-', ''))
+        self.persist_temp_table = persist_temp_table
+        self.delegate_to = delegate_to
+        self.labels = labels
+        self.location = location
+        self.default_table_expiry_in_ms = default_table_expiry_in_ms
+
+    def execute(self, context):
+        result_export_success = True
+        dataset_creation_success = False
+        query_execution_success = False
+        err_msg = ""
+        try:
+            hook = BigQueryHook(use_legacy_sql=False,
+                                bigquery_conn_id=self.bigquery_conn_id,
+                                delegate_to=self.delegate_to,
+                                location=self.location)
+            service = hook.get_service()
+            cursor = BigQueryBaseCursor(project_id=self.project_id,
+                                        service=service)
+            cursor.create_empty_dataset(dataset_id=self.dataset_id,
+                                        project_id=self.project_id,
+                                        dataset_reference={'defaultTableExpirationMs': self.default_table_expiry_in_ms})
+            dataset_creation_success = True
+            cursor.run_query(destination_dataset_table=self.temp_table_name,
+                             write_disposition='WRITE_TRUNCATE',
+                             allow_large_results=True,
+                             sql=self.query,
+                             use_legacy_sql=False)
+            query_execution_success = True
+            cursor.run_extract(source_project_dataset_table=self.temp_table_name,
+                               destination_cloud_storage_uris=self.destination_cloud_storage_uris,
+                               compression=self.compression,
+                               export_format='CSV',
```

Comment:
Why CSV?
BigQuery supports Arrays / Record types. These can not be converted to CSV.

Reply:
I fixed this